src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,157 @@
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Malte Grave - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+
+#include <cstdint>
+#include <cstring>
+#include <fcntl.h>
+#include <iostream>
+#include <linux/can.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+#include "CANLayer.h"
+#include "basecommfb.h"
+#include "comtypes.h"
+#include "datatype.h"
+#include "devlog.h"
+#include "forte_any.h"
+#include "forte_byte.h"
+
+using namespace forte::com_infra;
+
+CCANComLayer::CCANComLayer(CComLayer *paUpperLayer, CBaseCommFB *paComFB) :
+	CComLayer(paUpperLayer, paComFB) {}
+
+EComResponse CCANComLayer::openConnection(char *paLayerParameter) {
+	EComResponse eRetVal = e_InitInvalidId;
+	CParameterParser parser(paLayerParameter, ';', eCANComParamterAmount);
+	if (eCANComParamterAmount != parser.parseParameters()) {
+		return eRetVal;
+	}
+
+	setupCANInternals(parser);
+
+	switch (mFb->getComServiceType()) {
+	case e_Server:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Publisher block to send data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Client:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Subscribe block to receive data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Publisher:
+		// is handled via sendData
+		break;
+	case e_Subscriber:
+		// is handled via recvData
+		break;
+	}
+
+	DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+	CFDSelectHandler::TFileDescriptor fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+	if (CFDSelectHandler::scmInvalidFileDescriptor == fd) {
+		DEVLOG_DEBUG("[CAN Layer] Socket can't be created.\n");
+		return forte::com_infra::e_InitInvalidId;
+	}
+
+	mCANSocket = fd;
+	std::strcpy(mIfr.ifr_name, mCANParam.interfaceName.c_str());
+	ioctl(mCANSocket, SIOCGIFINDEX, &mIfr);
+	mAddr.can_family = AF_CAN;
+	mAddr.can_ifindex = mIfr.ifr_ifindex;
+
+	if (bind(mCANSocket, (struct sockaddr *)&mAddr, sizeof(mAddr)) < 0) {
+		DEVLOG_DEBUG("[CAN Layer] Binding of the socket unsuccessful.\n");
+		return eRetVal;
+	}
+
+	this->getExtEvHandler<CFDSelectHandler>().addComCallback(mCANSocket, this);
+	DEVLOG_DEBUG("[CAN Layer] Callback handler added.\n");
+	eRetVal = e_InitOk;
+
+	return eRetVal;
+}
+
+CCANComLayer::~CCANComLayer() { closeConnection(); }
+
+void CCANComLayer::closeConnection() {
+	if (mCANSocket != CFDSelectHandler::scmInvalidFileDescriptor) {
+		DEVLOG_DEBUG("[CAN Layer] Closing socket\n");
+		close(mCANSocket);
+	}
+	DEVLOG_DEBUG("[CAN Layer] Unregister layer\n");
+	this->getExtEvHandler<CFDSelectHandler>().removeComCallback(mCANSocket);
```

Comment:
probably this needs to go inside the `if` as the callback was not added if the socket is invalid

Reply:
I'm not sure about this one. I'm return in each error case, so this would be okay. Or am I missing something?

Reply:
Ah sorry wrong line :D